### PR TITLE
Refactor GMRF and remove physical dim parameter

### LINKED ISSUE #158: Bug: GMRF initializer applies the method mean.reshape  even if the mean is a model, need to fix this or generate proper error message.
Perhaps also switch precision parameter to scale.

Look at the code for LMRF and CMRF. Use similar init strategy and overvall argument handling.

Also look at how it handles init with geometries. You should be able to give Image2D geometry as inputl (remove physical dim)

may be related to #255

### PR DESCRIPTION
**Pull request by nabriis**
Closes #158, #255

This refactor also uncovered a very subtle bug with conjugate relation between GMRF and Gamma. There was a bug in the code, where the conjugate update for GMRF-Gamma pair was wrong because the mean of GMRF was assumed to be (dim,) array, but it was (dim, 1) array. This is now fixed and this improved the hierachical models with GMRF a lot. 

Here is how the  results for Paper 1 hierachical problem:
```python
A, y_data, info = Deconvolution1D(phantom='sinc', noise_std=0.01).get_components()
n = A.domain_dim

s = Gamma(1, 1e-4)
d = Gamma(1, 1e-4)
x = GMRF(np.zeros(n), prec = lambda d: d)
y = Gaussian(A@x, prec = lambda s: s)

BP = cuqi.problem.BayesianProblem(y, x, d, s)
BP.set_data(y=y_data)

BP.UQ(1000, exact={"x": x_true, "s": 1/0.01**2})
```
would have looked like before (with gamma hyperprior on GMRF),
image

and here is how it looks now:
image

Decided not to change the parameter to e.g. `location, scale` or `mean, cov`.

Regardless of the choice it will either be inconsistant with other MRFs or Gaussian. And Prec is the actual main parameter that is used when working with conjugate priors, so I kept it like that.

### CODE REVIEW

**Review by jakobsj — APPROVED** [review]
From a brief look, LGTM.

**Comment by jakobsj** [comment]
Could we possibly allow passing a scalar for the mean/location?  The typical case is passing a zero, and with passing a zero vector of size A.domain_dim looks a bit clunky, especially since we also pass the A.domain_geometry, which appears a little bit redundant.  In particular, this impacts the cookie case and CT cases which we will benefit from looking as simple as possible.

**Comment by jakobsj** [comment]
That comment is for all of GMRF, LMRF and CMRF. It makes a big simplification in the code snippets, if one can pass just 0 instead of np.zeros(A.domain_dim).....  @nabriis Will this be possible?

**Comment by nabriis** [comment]
> Could we possibly allow passing a scalar for the mean/location? The typical case is passing a zero, and with passing a zero vector of size A.domain_dim looks a bit clunky, especially since we also pass the A.domain_geometry, which appears a little bit redundant. In particular, this impacts the cookie case and CT cases which we will benefit from looking as simple as possible.

Great question @jakobsj. This should indeed be possible for all of CMRF, GMRF, LMRF. This is tested via the code in https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_distributions_shape.py where GMRF is not included as part of this PR.

**Comment by jakobsj** [comment]
Thanks, that sounds good!   I don't follow exactly what you say though, are you saying it is not currently possible in the existing code, also not with the present PR, but should be simple to add?  If so, could you add it for GMRF in this PR?  ANd a separate PR for LMRF and CMRF?   It seems a simple check, if given a scalar, check the dimension in the geometry and expand the scalar to vector of that size?

Semi-related, will it now be mandatory to pass a geometry for G/L/CMRF?  Currently some paper code snippets contain (for a 1D case) something like GMRF(xzero, 10), where xzero is a np.zeros vector from which the size is then determined. Should this be replaced by passing the 1D geometry or do we want this syntax to still be allowed and use a (defaultgeometry?)  default 1D geometry?

**Comment by nabriis** [comment]
> Thanks, that sounds good! I don't follow exactly what you say though, are you saying it is not currently possible in the existing code, also not with the present PR, but should be simple to add? If so, could you add it for GMRF in this PR? ANd a separate PR for LMRF and CMRF? It seems a simple check, if given a scalar, check the dimension in the geometry and expand the scalar to vector of that size?
> 
> Semi-related, will it now be mandatory to pass a geometry for G/L/CMRF? Currently some paper code snippets contain (for a 1D case) something like GMRF(xzero, 10), where xzero is a np.zeros vector from which the size is then determined. Should this be replaced by passing the 1D geometry or do we want this syntax to still be allowed and use a (defaultgeometry?) default 1D geometry?

It is possible for LMRF and CMRF to pass scalar inputs and a geometry in main. After this PR, it will also be possible for GMRF. It was not possible before this PR. Note: It is actually not as simple as to just check during initialization since we allow all distribution inputs to be callable functions, meaning it needs to be handled in a different way :) Anyway, its tested automatically and works.

As for the question on whether geometry is mandatory. In main we allow LMRF and CMRF to be defined with either a geometry or by passing a vector as the location parameter (where we then set a default geometry of the same shape). This PR adds this behavior to GMRF.

**Comment by jakobsj** [comment]
That sounds great, thank you for the clarifications.  I will then proceed
with scalar input notation to keep code snippets concise.  And probably
stick with passing the xzero vector input.

File: cuqi/distribution/_gmrf.py [hunk 1]
```diff
@@ -94,54 +95,67 @@ class GMRF(Distribution):
     For more details see: See Bardsley, J. (2018). Computational Uncertainty Quantification for Inverse Problems, Chapter 4.2.
 
     """
-        
-    def __init__(self, mean, prec, physical_dim=1, bc_type='zero', order=1, is_symmetric=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/distribution/_gmrf.py:98]: Thanks! Missed that.

**Pull request merged**